### PR TITLE
Fix AgentRunner AgentRunStepStartEvent dispatch

### LINKED ISSUE #14822: [Bug]: AgentRunStepStartEvent may be dispatched without associated TaskStep
### Bug Description

At the very beginning of `AgentRunner._run_step`, an `AgentRunStepStartEvent` is dispatched. In the current implementation, the value of the `TaskStep` object attached to `AgentRunStepStartEvent` will be`None` if the step is popped from `step_queue` (as opposed to being passed explicitly as an argument to `AgentRunner._run_step`). This means that in many cases, the `TaskStep` associated with the current step being executed is not attached to the corresponding `AgentRunStepStartEvent`.

> 

### Version

0.10.55

### Steps to Reproduce

1. Attach an event handler to the root dispatcher to record / inspect events
2. Create an agent that will run one or more steps
3. Run the agent
4. Inspect all `AgentRunStepStartEvent` dispatched, and notice that their `step` field is `None` despite the fact that one or more steps were executed

### Relevant Logs/Tracbacks

_No response_

### PR DESCRIPTION
**Pull request by joelrorseth**
# Description

Fixing an issue where in many cases, the `AgentRunStepStartEvent` dispatched in `AgentRunner._run_step` is not assigned the current `TaskStep`. I added an extensive test that verifies all agent-related events being dispatched.

Fixes #14822 

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-core/tests/agent/runner/test_base.py [hunk 1]
```diff
@@ -266,3 +287,61 @@ def test_agent_from_llm() -> None:
...
+        assert event_handler.events[run_step_start_idx].step is not None
+        assert (
+            event_handler.events[run_step_start_idx].step.task_id
+            == event_handler.events[run_step_start_idx].task_id
+        )
+        assert event_handler.events[run_step_start_idx].input is None
+
+    # Check all AgentRunStepEndEvent
+    for i in range(num_steps):
+        run_step_start_idx = 2 * i + 2
```
Comment (nerdai) [inline llama-index-core/tests/agent/runner/test_base.py:323]: nit: should this be `run_step_end_idx`?

**Review by nerdai — APPROVED** [review]
Looks good @joelrorseth!

Left a minor nit comment in my review.

File: llama-index-core/tests/agent/runner/test_base.py (hunk 1 above)
Comment (joelrorseth) [reply llama-index-core/tests/agent/runner/test_base.py]: Oops, good catch, fixed!

Comment (nerdai) [reply llama-index-core/tests/agent/runner/test_base.py]: lgtm! thanks :)

**Pull request merged**